Issue 2772 - part C: externalize with majority signed

hewison-chris opened the pull request:
Once the `valueExternalized` single SCP event occurs for a slot we sign and gossip our signature for this `pending block`. Each time we add signatures for this block in `updateMultiSignature` we check if we have enough for a majority to add the block to the ledger. Any SCP envelopes received whilst waiting will be stored to be processed once the pending block is finalized (written to the ledger).

hewison-chris committed:
Trivial correction of indentation

hewison-chris committed:
Externalize blocks only with majority signatures

When the SCP `Externalize` event occurs then the node will wait for the
majority of signatures before writing the block to the ledger.

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -769,6 +772,20 @@ extern(D):
+
+        if (block_sig.height == pending_block.header.height)
+        {
+            if (this.collectBlockSignature(block_sig, pending_block.hashFull()))
+            {
+                log.dbg("{}: Add to next block #{} being externalized for utxo {}", __PRETTY_FUNCTION__, block_sig.height, block_sig.utxo);
+                log.dbg("{}: header was {}", __FUNCTION__, pending_block.header.validators);
+                this.updateMultiSignature(pending_block.header);
+                log.dbg("{}: header now {}", __FUNCTION__, pending_block.header.validators);
+            }
```
The calls of the `collectBlockSignature` and `updateMultiSignature` are shown twice in this function. Can't we refactor the code like this which is not complete code?
```
if (block_sig.height == pending_block.header.height)
    block = this.pending_block;
else
    block = this.ledger.getBlocksFrom(Height(block_sig.height)).front;

if (this.collectBlockSignature(block_sig, block.hashFull()))
{
    log.dbg("{}: Add to next block #{} being externalized for utxo {}", __PRETTY_FUNCTION__, block_sig.height, block_sig.utxo);
    log.dbg("{}: header was {}", __FUNCTION__, block.header.validators);
    this.updateMultiSignature(block.header);
    log.dbg("{}: header now {}", __FUNCTION__, block.header.validators);
}
else
{
    return BlockHeight.init;
}

if (block_sig.height == this.pending_block.header.height)
    return BlockHeight.init;
```

hewison-chris replied:
Refactored to be cleaner.

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1062,6 +1053,45 @@ extern(D):
+                    log.dbg("{}: Will not yet externalize block #{} as we only have {} / {} signatures",
+                        __FUNCTION__, pending_block.header.height, pending_block.header.validators.setCount,
+                        pending_block.header.validators.count);
+                    this.timers[TimersIdx.Catchup].rearm(CatchupTaskDelay, false);
```
Use this.armTaskTimer please.

linked0 replied:
Ah, I'm sorry. This could have to remain as it is.

hewison-chris replied:
Updated

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1062,6 +1053,45 @@ extern(D):
+                        __FUNCTION__, pending_block.header.height, pending_block.header.validators.setCount,
+                        pending_block.header.validators.count);
+                    this.timers[TimersIdx.Catchup].rearm(CatchupTaskDelay, false);
+                    this.timers[TimersIdx.Externalize].rearm(ExternalizeTaskDelay, false);
```
Ditto

hewison-chris replied:
Updated

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1062,6 +1053,45 @@ extern(D):
+            }
+            pending_block = Block.init;
+            // Enable envelope processing again
+            this.timers[TimersIdx.Envelope].rearm(EnvTaskDelay, false);
```
Ditto

hewison-chris replied:
Updated

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -647,7 +646,11 @@ extern(D):
             else
                 this.queued_envelopes.insertFront(copied);
-            this.armTaskTimer(TimersIdx.Envelope, EnvTaskDelay);
+            if (pending_block == Block.init)
```
```suggestion
            if (this.pending_block == Block.init)
```

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -769,6 +772,20 @@ extern(D):
         log.trace("Received BLOCK SIG {} from node {} for block {}",
                     block_sig.signature, block_sig.utxo, block_sig.height);
+
+        if (block_sig.height == pending_block.header.height)
```
```suggestion
        if (block_sig.height == this.pending_block.header.height)
```

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1042,16 +1035,14 @@ extern(D):
             }
 
-            Block block = this.ledger.buildBlock(
+            pending_block = this.ledger.buildBlock(
```
```suggestion
            this.pending_block = this.ledger.buildBlock(
```

hewison-chris replied:
Updated

linked0 reviewed (commented):
Very nice! Some suggestions.

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1062,6 +1053,45 @@ extern(D):
+                    this.timers[TimersIdx.Externalize].rearm(ExternalizeTaskDelay, false);
+                    return;
+                }
+                else
```
Isn't there any possibility of `pending_block.header.height > this.ledger.height + 1` like in the middle of catch-up?

hewison-chris replied:
No as we check in `valueExternalized` the current ledger height before creating a `pending_block`.

hewison-chris committed:
Check if shutting down before running tasks

It can happen that the node is shutting down when `pushBlock`,
`pushBlockHeader`, `pushTransaction` or `pushPreimage` are called. If
this happens then the handler array might have been set to  `null`.

omerfirmak commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -647,7 +646,11 @@ extern(D):
+            if (this.pending_block == Block.init)
+                this.armTaskTimer(TimersIdx.Envelope, EnvTaskDelay);
+            else
+                log.dbg("{}: We have a pending block #{} to externalize so do not process any more envelopes yet",
```
This does not quite guarantee that we wont process them tho, envelope task could be already running or armed. So a similar check should be put in `envelopeProcessTask` too

hewison-chris replied:
Good point, I'll add a check.

hewison-chris replied:
Updated

linked0 reviewed (approved):
LGTM

omerfirmak commented on source/agora/consensus/state/Ledger.d:
```diff
@@ -350,10 +350,6 @@ public class Ledger
                 ? null
                 : "Trying to externalize a different block for the latest height";
-
-        if (!this.hasMajoritySignature(this.height()))
```
should we keep these as an assertion?

hewison-chris replied:
Updated

omerfirmak commented on source/agora/node/Validator.d:
```diff
@@ -99,9 +99,6 @@ public class Validator : FullNode, API
     private Identity identity;
 
-    /// Signatures we received a early
-    private Set!ValidatorBlockSig early_sigs;
```
These helped with signatures arriving before SCP signals us the externalization event. I think that is still possible even if we delay the actual externalization, no?

hewison-chris replied:
Yes that is true. I'll revert this change and store these early signatures.

hewison-chris replied:
Moved them to `Nominator` and store them until block is pending.

omerfirmak commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1062,6 +1055,45 @@ extern(D):
+    }
+
+    /// If we have majority signatures then externalize to the ledger otherwise try again later
+    extern(D) private void checkExternalize () nothrow
```
I dont think this task is really necessary. I feel like we can use `updateMultiSignature` as a checkpoint to determine if we can externalize the pending block.

hewison-chris replied:
Good idea, I'll give it a go.

hewison-chris replied:
Updated.

omerfirmak commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1042,16 +1036,17 @@ extern(D):
+            this.slot_sigs[height][self] = this.signBlock(this.pending_block);
+            this.updateMultiSignature(this.pending_block.header);
+            // Now add any early signatures we received
+            early_sigs[].each!(sig => this.receiveBlockSignature(sig));
```
We wont gossip the early sigs this way tho, right?

hewison-chris replied:
Yes it will get gossiped in `updateMultisignature` when it is handled again in `receiveBlockSignature`.

omerfirmak commented:
Overall LGTM apart from the last comment I left, but I would say we can wait for another set of eyes on this since it is pretty important that we get this mostly right :)

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -924,12 +842,7 @@ public class ValidatingLedger : NodeLedger
                 data.enrolls ~= enroll;
             }
         }
-
-        if (auto reason = this.externalize(data, sign_all))
-        {
-            assert(0, format!"Failure in unit test. Block %s should have been externalized: %s"(
-                       this.height() + 1, reason));
-        }
```
This change seems unrelated, and I don't think we want it. `forceCreateBlock` should not fail (it's a test-only utility).

hewison-chris replied:
Reverted

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -1791,45 +1704,6 @@ unittest
     scope ledger = new TestLedger(WK.Keys.A, [GenesisBlock], params);
 }
 
-unittest
```
Why is this test removed ?

hewison-chris replied:
It was testing that when previous block has less than majority the next block is not externalized. Maybe I should bring it back but to test the current block is not externalized with less than majority.

hewison-chris replied:
Added

Geod24 commented on source/agora/consensus/state/Ledger.d:
```diff
@@ -402,8 +402,8 @@ public class Ledger
-        log.info("Beginning externalization of block #{} (previous block signatures: {})",
-            block.header.height, this.last_block.header.validators);
+        log.info("Beginning externalization of block #{} (block signatures: {})",
+            block.header.height, block.header.validators);
```
This is already printed later, let's keep the previous block signatures here.

hewison-chris replied:
Was thinking it was not so interesting now it does not block the next block. However it is still useful info so will revert.

hewison-chris replied:
Reverted

Geod24 commented on source/agora/consensus/state/Ledger.d:
```diff
@@ -436,7 +436,8 @@ public class Ledger
         this.last_block = deserializeFull!Block(serializeFull(block));
-        log.info("Completed externalization of block #{}", block.header.height);
+        log.info("Completed externalization of block #{} (block signatures: {})",
+            block.header.height, block.header.validators);
```
Printed L409

hewison-chris replied:
Reverted

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -652,6 +645,9 @@ public class Validator : FullNode, API
     private void buildRequiredKeys (NodeID self_id, in QuorumConfig quorum_conf) @safe
     {
         auto validators = this.getValidators();
+        assert(quorum_conf.nodes.length <= validators.length,
+            format!"Too many quorum nodes for next block #%s (%s > %s)"
+                (this.ledger.height() + 1, quorum_conf.nodes.length, validators.length));
```
That looks unrelated ?

hewison-chris replied:
Extracted into separate commit

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -618,6 +610,7 @@ public class Validator : FullNode, API
             // Try to register immediately if we are a validator for the next cycle
             if (this.enroll_man.isEnrolled(block.header.height + 1, &this.ledger.peekUTXO))
                 this.network.onRegisterName();
+            assert(block.header.height <= this.ledger.height);
```
Why?

hewison-chris replied:
Extracted into separate commit.

Geod24 commented on source/agora/consensus/state/Ledger.d:
```diff
@@ -899,35 +900,12 @@ public class Ledger
 
-    protected bool hasMajoritySignature (in BlockHeader header,
-        in ValidatorInfo[] validators) @safe nothrow
+   public bool hasMajoritySignature (in BlockHeader header) @safe nothrow
```
`nothrow @nogc` too right ?

hewison-chris replied:
Actually not easy as `percentage` of `BitMask` is not `@nogc` due to `notSetIndices` using a lambda.

Geod24 commented on source/agora/common/BitMask.d:
```diff
@@ -130,6 +130,7 @@ public struct BitMask
     /// return the percentage of bits set to `true`
     public ubyte percentage () const
     {
+        assert(this.length > 0);
```
Own commit

hewison-chris replied:
Extracted

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1127,6 +1160,9 @@ extern(D):
         header.updateSignature(multiSigCombine(sigs_to_add), validator_mask);
         log.trace("Updated block signature for block {}, mask: {}",
                 header.height, validator_mask);
+        log.dbg("{}: Number of signatures in header for block #{} is now {}",
+            __FUNCTION__, header.height, header.validators.setCount);
```
Isn't it redundant with the `trace` message just before ?

hewison-chris replied:
I'll combine the info into the Trace message.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1064,8 +1059,44 @@ extern(D):
+        {
+            if (block.header.height == this.ledger.height + 1)
+            {
+                if (this.ledger.hasMajoritySignature(block.header))
```
At this point we should just remove `hasMajoritySignature` and check `percentage`.

hewison-chris replied:
In `Ledger.hasMajoritySignature` it does exactly that. Or do you want to avoid the function call?

Geod24 replied:
Well it's not a property of the `Ledger` anymore: We use it on non-externalize blocks, and it doesn't depend on `ConsensusParams`, so I don't know why it should stay in the `Ledger`.

hewison-chris replied:
Updated

hewison-chris committed:
Assert that quorum nodes is not more than validators

When building required peers after quorum shuffle we must not have more
quorum nodes in the configuration than validators.

hewison-chris committed:
Assert correct block height used to regenerate quorums

When a block is added to the ledger and we call `regenerateQuorums` the
block height passed in must be the ledger block height.

hewison-chris committed:
Trivial function documentation update

hewison-chris committed:
Assert divisor is greater than zero

When we calculate the percentage of validators who signed the number of
validators must be more than zero.

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -1800,34 +1713,14 @@ unittest
-    // should be able to create a new block after previous height receives majority signatures
-    ledger.forceCreateBlock();
-    assert(ledger.height() == 2);
+    // if at least 4 out of 6 validators sign it should externlize
+    iota(4, 7).each!(signed =>
+        assert(!ledger.externalize(ConsensusData.init, signed)));
```
Don't we externalize 3 blocks in a row here ?

hewison-chris replied:
Yes it would. Shall I add a ledger height assertion?

hewison-chris replied:
Added the height assertions.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -769,16 +769,34 @@ extern(D):
+
+        if (!pendingBlockSig && block_sig.height > cur_height) // Too early for this signature
+        {
+            early_sigs.put(block_sig);
```
Hold on, if someone sends a fake signature for 100 blocks in the future, we will always store it, won't we ?

hewison-chris replied:
Good point, I will add a limit of 3 blocks.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -769,16 +769,34 @@ extern(D):
+            return BlockHeader.init; // We want the caller to ignore this signature
+        }
+
+        Block block = pendingBlockSig ? this.pending_block : this.ledger.getBlocksFrom(Height(block_sig.height)).front;
```
Also if someone sends a signature for block `0` while we don't have a pending block signature, we will store it, won't we ?

hewison-chris replied:
Added an early return for this condition of a block signature for height 0 or for the case where the signature is for more than 3 blocks into the future.

Geod24 reviewed (approved):
Mostly LGTM, just a couple questions

omerfirmak commented:
Next step IMO is to put a timeout on how long a pending block can exist and if we reach it, start a new consensus round. so our assumption `block_height == slot_idx` won't hold anymore.

hewison-chris commented:
> Next step IMO is to put a timeout on how long a pending block can exist and if we reach it, start a new consensus round. so our assumption `block_height == slot_idx` won't hold anymore.

Having the `block_height != slot_idx` would be very stressful in my opinion.
